services/server/src/server/services/workers/verificationWorker.ts

```diff
@@ -0,0 +1,158 @@
+import { VerifyFromJsonInputs, VerifyFromJsonOutput } from "./workerTypes";
+
+export const filename = resolve(__filename);
+
+let chainRepository: ChainRepository;
+let solc: SolcLocal;
+let vyper: VyperLocal;
+
+const initWorker = () => {
+  if (chainRepository && solc && vyper) {
+    return;
+  }
+
+  const sourcifyChainInstanceMap = Piscina.workerData
+    .sourcifyChainInstanceMap as { [chainId: string]: SourcifyChainInstance };
+
+  const sourcifyChainMap = Object.entries(sourcifyChainInstanceMap).reduce(
+    (acc, [chainId, chain]) => {
+      acc[chainId] = new SourcifyChain(chain);
+      return acc;
+    },
+    {} as SourcifyChainMap,
+  );
+
+  chainRepository = new ChainRepository(sourcifyChainMap);
+  solc = new SolcLocal(
+    Piscina.workerData.solcRepoPath,
+    Piscina.workerData.solJsonRepoPath,
+  );
+  vyper = new VyperLocal(Piscina.workerData.vyperRepoPath);
+};
+
+export async function verifyFromJsonInput({
+  chainId,
+  address,
+  jsonInput,
+  compilerVersion,
+  contractIdentifier,
+  creationTransactionHash,
+}: VerifyFromJsonInputs): Promise<VerifyFromJsonOutput> {
+  initWorker();
+
+  // The contract path can include a colon itself,
+  // therefore we need to take the last element as the contract name
+  const splitIdentifier = contractIdentifier.split(":");
+  const contractName = splitIdentifier[splitIdentifier.length - 1];
+  const contractPath = splitIdentifier.slice(0, -1).join(":");
+  const compilationTarget: CompilationTarget = {
+    name: contractName,
+    path: contractPath,
+  };
+
+  const sourcifyChain = chainRepository.sourcifyChainMap[chainId];
+
+  const foundCreationTxHash =
+    creationTransactionHash ||
+    (await getCreatorTx(sourcifyChain, address)) ||
+    undefined;
+
+  let compilation: SolidityCompilation | VyperCompilation | undefined;
+  try {
+    if (jsonInput.language === "Solidity") {
+      compilation = new SolidityCompilation(
+        solc,
+        compilerVersion,
+        jsonInput as SolidityJsonInput,
+        compilationTarget,
+      );
+    } else if (jsonInput.language === "Vyper") {
+      compilation = new VyperCompilation(
+        vyper,
+        compilerVersion,
+        jsonInput as VyperJsonInput,
+        compilationTarget,
+      );
+    }
+  } catch (error: any) {
+    return {
+      errorResponse: createMatchingError(error),
+    };
+  }
+
+  if (!compilation) {
+    return {
+      errorResponse: {
+        customCode: "unsupported_language",
+        message: getVerificationErrorMessage("unsupported_language"),
+        errorId: uuidv4(),
+      },
+    };
+  }
+
+  const verification = new Verification(
+    compilation,
+    sourcifyChain,
+    address,
+    foundCreationTxHash,
+  );
+
+  try {
+    await verification.verify();
+  } catch (error: any) {
+    return {
+      errorResponse: createMatchingError(error, verification),
+    };
+  }
+
+  return {
+    verificationExport: verification.export(),
+  };
```

Comment:
Can we pass directly the export? Why do we need an object?

Reply:
I was actually doing this earlier, but realized then that we also need to return errors in the response. Potentially, you could just throw errors, but the structured clone algorithm would only copy the default properties (name and message) of the Error class back to the main thread. We also need to pass our custom code and bytecodes back. So I needed the response object to be able to also return an error (instead of throwing).

Reply:
Ok that's clear, maybe let's add a comment to explain that